### PR TITLE
LPS-113302 Apply classic ckeditor in Message Boards, part 2/2

### PR DESCRIPTION
**Pull request by markocikos**
@jbalsas @wincent @carloslancha 

In this PR, we are updating bbcode editor in message boards, fixing a regression bug and cleaning up the `ClassicEditor` a bit.

To test:

1. Site Administatration > Content & Data > Message Boards > options on top right > Configuration
1. Set format to BBCode
1. Place the Message Boards App on a page
1. New Thread > Make a sample post. Add some sample formatting, like bolding. Publish.
1. Edit post.
1. Cancel.
1. Reply.
1. Cancel reply.
1. Reply.
1. Post reply.

Notes:
1. We are not building a new editor for bbcode encoding. Instead, we are using `ckeditor_classic`, just with different CKEditor configuration. This is not a scary change, since the only use of bbcode encoding is in message boards. There is a mention in `mentions-web`, but no `<liferay-editor:editor>`, so not sure if or how it applies.
1. Instead of setting editor content with `data` prop, we are setting data on `instanceReady`. This way plugins are applied before setting data, including encoding plugin `bbcode`. Without this change, editor shows bbcode source code when it should show executed HTML. Alternative would be to reformat data before passing it editor, like we are doing on [view message page](https://github.com/liferay/liferay-portal/blob/74e300f3f9a09543a8c6521c1c9e05d7e9a2d8e0/modules/apps/message-boards/message-boards-web/src/main/resources/META-INF/resources/message_boards/view_thread_message.jsp#L403), but in my opinion it's better to use editor's own API. We end up using it anyways, for example on switching to Source and back.
1. I removed `cssClass` prop as we are not using it. There is the `cssClasses` attribute [floating around backend configuration](https://github.com/liferay/liferay-portal/blob/74e300f3f9a09543a8c6521c1c9e05d7e9a2d8e0/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/java/com/liferay/frontend/editor/ckeditor/web/internal/editor/configuration/CKEditorClassicConfigContributor.java#L69), but it is added to editor config object.
1. I removed the `key` prop, as far as I can tell it is unnecessary.
1. The regression bug is on opening a reply post and canceling it. It happens because we didn't migrate `window[editorName].dispose` to `ClassicEditor`. This is the only use of `dispose`. Instead of migrating it now, in this PR I opted to remove the usage. Now, when we hide the reply post, editor is just hidden with it. Replying again re-initializes editor with no problem.

---

In the legacy editor, as well as partially migrated to the new component, there is this rather ugly pattern of defining utility methods in the `window[editorName]`. The mentioned `dispose` is one example, but `getHTML` is used a lot more:
```js
useEffect(() => {
	window[name] = {
		getHTML,
		getText() {
			return contents;
		},
	};
}, [contents, getHTML, name]);
```
To replace it, we could make a new utility, something like `Liferay.CKEditor.getData(name)`, that would serve the same purpose. The content of the utility would be basically
```js
data = editor.getData();

if (CKEDITOR.env.gecko && CKEDITOR.tools.trim(data) === '<br />') {
    return '';
}

return data;
```
What do you think?

### CODE REVIEW

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
ci:test:relevant

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js [hunk 1]
```diff
@@ -100,22 +99,24 @@ const ClassicEditor = ({
...
-				key={toolbarSet}
 				onBeforeLoad={(CKEDITOR) => {
 					CKEDITOR.disableAutoInline = true;
 					CKEDITOR.dtd.$removeEmpty.i = 0;
 					CKEDITOR.dtd.$removeEmpty.span = 0;
 
 					CKEDITOR.on('instanceCreated', ({editor}) => {
 						editor.name = name;
+
+						editor.on('instanceReady', () => {
```
Comment (jbalsas) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js:117]: I remember we had some `processorsLoaded` event in the past... I'm assuming that's properly fired inside `instanceReady`?

**Comment by jbalsas** [comment]
If tests are green, I see no reason to not forward this :)

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
> In the legacy editor, as well as partially migrated to the new component, there is this rather ugly pattern of defining utility methods in the `window[editorName]`. The mentioned `dispose` is one example, but `getHTML` is used a lot more:
> 
> ```js
> useEffect(() => {
> 	window[name] = {
> 		getHTML,
> 		getText() {
> 			return contents;
> 		},
> 	};
> }, [contents, getHTML, name]);
> ```
> 
> To replace it, we could make a new utility, something like `Liferay.CKEditor.getData(name)`, that would serve the same purpose. The content of the utility would be basically
> 
> ```js
> data = editor.getData();
> 
> if (CKEDITOR.env.gecko && CKEDITOR.tools.trim(data) === '<br />') {
>     return '';
> }
> 
> return data;
> ```
> 
> What do you think?

Well, the think is, we still support other editors and `window[editorName]` provides an abstraction layer. I don't see how `Liferay.Editor.getData(...)` would actually be better. Wouldn't you still need to wrap those React components or register them somewhere to get access to the content?

It's been my dream forever to avoid this programatic API, though 😂 

We have several usages that I know of for such an API:
- Form validation. Triggered by aui-validator, we will call the `getHTML` method to validate its contents. 
- Form submission. When submitting a form, we often need to retrieve the contents of the Editor and place them in a given `input` element
- Other fancy stuff :)

My suggestion would be to see if we can start removing some of those needs. For example, automatically set the value to the input on `Liferay.on('submit')` so we can remove the ones that use `getHTML` for that... then see options for the rest of usages and so on...

**Comment by markocikos** [comment]
> Well, the think is, we still support other editors and `window[editorName]` provides an abstraction layer. I don't see how `Liferay.Editor.getData(...)` would actually be better. Wouldn't you still need to wrap those React components or register them somewhere to get access to the content?

I see these benefits of `Liferay.Editor.getData`
1. Readability. When you see something like `window['<%= name %>'].getHTML()` or `<portlet:namespace />getHTML()` in code, it's not clear what's going on. You're getting HTML, that's clear, but from what?
1. Consistency. This is basically a utility that is a thin wrapper around CKEditor's `getData()`. A standard way of implementing such a utility would be `getEditorData` for ES6 context and `Liferay.Editor.getData` for other contexts.
1. Memory footprint. When we have several instances of editor on a page, each one of them would generate it's own API. Instead, with a standard util, we have one API that can access an instance based on argument. This is probably a minor benefit, I don't think there are many pages with multiple editors.

Of course, removing the need to use such a utility, as you are suggesting, would be even better!

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Let's try again... maybe one day we'll see it green... 

/cc @john-co, @kenjiheigel

**Comment by jbalsas** [comment]
ci:test:relevant

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js:117]: We are still firing [`customDataProcessorLoaded`](https://github.com/liferay/liferay-portal/blob/579f29bf87652eab0e97183802a6374708e0698c/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/bbcode/bbcode_data_processor.js#L961), but we are no longer using it. Its only listener is in ckeditor.jsp. I guess that is an alternative solution for timing when the data can be set.

**Comment by jbalsas** [comment]
hallelluja

**Comment by jbalsas** [comment]
ci:forward

**Comment by john-co** [comment]
All tests passed but the top level build failed

**Error Snippet:**
```[beanshell] java.util.concurrent.TimeoutException: Unable to find downstream build. Invocation URL:https://test-1-12.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/buildWithParameters?DIST_PATH=/tmp/dist/test-1-18/test-portal-acceptance-pullrequest(master)/3994/dist&JENKINS_GITHUB_PULL_REQUEST_NUMBER=0&JOB_VARIANT=functional-upgrade-tomcat90-mysql57-jdk8/3&JSON_MAP_URL=http://cloud-10-0-0-31/osb-jenkins-web/map/test-1-18/test-portal-acceptance-pullrequest(master)/3994/test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8/3/start.properties&BUILD_PRIORITY=5&DIST_NODES=cloud-10-0-16-20,cloud-10-0-110-197,cloud-10-0-19-114,cloud-10-0-16-148,cloud-10-0-12-76,cloud-10-0-19-25,cloud-10-0-16-191,cloud-10-0-16-170,cloud-10-0-19-231,cloud-10-0-10-57,cloud-10-0-13-124,cloud-10-0-13-120,cloud-10-0-12-68,cloud-10-0-13-181,cloud-10-0-110-250,cloud-10-0-13-55,cloud-10-0-19-86,cloud-10-0-19-155,cloud-10-0-110-18,cloud-10-0-110-177,cloud-10-0-13-110,cloud-10-0-19-221,cloud-10-0-110-176,cloud-10-0-10-47&JENKINS_GITHUB_SENDER_BRANCH_SHA=8082c2bc99d3f7da7225e7e21a90d582f2f9e637&JENKINS_GITHUB_UPSTREAM_BRANCH_NAME=master&AXIS_VARIABLES=0&JENKINS_GITHUB_UPSTREAM_BRANCH_SHA=8082c2bc99d3f7da7225e7e21a90d582f2f9e637&token=********
[beanshell] 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[beanshell] 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[beanshell] 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
```

cc/ @kenjiheigel, @yichenroy

**Pull request closed without merging**